satellite/gracefulexit/endpoint.go

```diff
@@ -221,15 +224,54 @@ func (endpoint *Endpoint) doProcess(stream processStream) (err error) {
 
 	for {
 		pendingCount := pending.length()
+
 		// if there are no more transfers and the pending queue is empty, send complete
 		if atomic.LoadInt32(&morePiecesFlag) == 0 && pendingCount == 0 {
-			// TODO check whether failure threshold is met before sending completed
 			// TODO needs exit signature
 			transferMsg := &pb.SatelliteMessage{
 				Message: &pb.SatelliteMessage_ExitCompleted{
 					ExitCompleted: &pb.ExitCompleted{},
 				},
 			}
+
+			exitStatusRequest := &overlay.ExitStatusRequest{
+				NodeID:         nodeID,
+				ExitSuccess:    true,
+				ExitFinishedAt: time.Now().UTC(),
+			}
+
+			progress, err := endpoint.db.GetProgress(ctx, nodeID)
+			if err != nil && !errs.Is(err, sql.ErrNoRows) {
+				return Error.Wrap(err)
+			}
+
+			// check node's exiting progress to see if it has failed passed max failure threshold
+			if progress.PiecesFailed > 0 {
+				exitStatusRequest.ExitSuccess = false
+				transferMsg.Message = &pb.SatelliteMessage_ExitFailed{
+					ExitFailed: &pb.ExitFailed{
+						Reason: pb.ExitFailed_OVERALL_FAILURE_PERCENTAGE_EXCEEDED,
+					},
+				}
+
+				if progress.PiecesTransferred > 0 {
+					overallFailurePercentage := math.Round(float64(progress.PiecesFailed)/float64(progress.PiecesTransferred)) * 100
+					if int(overallFailurePercentage) < endpoint.config.OverallMaxFailuresPercentage {
+						exitStatusRequest.ExitSuccess = true
+						transferMsg = &pb.SatelliteMessage{
+							Message: &pb.SatelliteMessage_ExitCompleted{
+								ExitCompleted: &pb.ExitCompleted{},
+							},
+						}
+					}
+				}
+			}
+
+			_, err = endpoint.overlaydb.UpdateExitStatus(ctx, exitStatusRequest)
+			if err != nil {
+				return Error.Wrap(err)
+			}
+
```

Comment:
I think we want to add `err := chore.db.DeleteTransferQueueItems(ctx, node.NodeID)` here as well.